File: java/src/test/java/io/cucumber/query/QueryAcceptanceTest.java
package io.cucumber.query;

import com.fasterxml.jackson.core.util.DefaultPrettyPrinter;
import io.cucumber.messages.Convertor;
import io.cucumber.messages.LocationComparator;
import io.cucumber.messages.NdjsonToMessageIterable;
import io.cucumber.messages.types.*;
import org.junit.jupiter.api.Disabled;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.MethodSource;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.Comparator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.function.Function;

import static com.fasterxml.jackson.core.util.DefaultIndenter.SYSTEM_LINEFEED_INSTANCE;
import static io.cucumber.query.Jackson.OBJECT_MAPPER;
import static io.cucumber.query.Repository.RepositoryFeature.INCLUDE_ATTACHMENTS;
import static io.cucumber.query.Repository.RepositoryFeature.INCLUDE_GHERKIN_DOCUMENTS;
import static io.cucumber.query.Repository.RepositoryFeature.INCLUDE_HOOKS;
import static io.cucumber.query.Repository.RepositoryFeature.INCLUDE_STEP_DEFINITIONS;
import static io.cucumber.query.Repository.RepositoryFeature.INCLUDE_SUGGESTIONS;
import static io.cucumber.query.Repository.RepositoryFeature.INCLUDE_UNDEFINED_PARAMETER_TYPES;
import static java.nio.charset.StandardCharsets.UTF_8;
import static java.util.Arrays.asList;
import static java.util.Comparator.nullsFirst;
import static java.util.stream.Collectors.toList;
import static org.assertj.core.api.Assertions.assertThat;

public class QueryAcceptanceTest {
    private static final NdjsonToMessageIterable.Deserializer deserializer = (json) -> OBJECT_MAPPER.readValue(json, Envelope.class);
    private static final Comparator<Pickle> pickleComparator = nullsFirst(Comparator.comparing(Pickle::getUri)
            .thenComparing(pickle -> pickle.getLocation().orElse(null), nullsFirst(new LocationComparator())));

    static List<QueryTestCase> acceptance() {
        List<QueryTestCase> testCases = new ArrayList<>();

        List<Path> sources = getSources();
        Map<String, Function<Query, Object>> queries = createQueries();

        sources.forEach(source ->
                queries.forEach((methodName, query) ->
                        testCases.add(new QueryTestCase(methodName, source, query))));

        return testCases;
    }

    private static List<Path> getSources() {
        return Arrays.asList(
                Paths.get("../testdata/src/attachments.ndjson"),
                Paths.get("../testdata/src/empty.ndjson"),
                Paths.get("../testdata/src/examples-tables.ndjson"),
                Paths.get("../testdata/src/global-hooks.ndjson"),
                Paths.get("../testdata/src/global-hooks-attachments.ndjson"),
                Paths.get("../testdata/src/hooks.ndjson"),
                Paths.get("../testdata/src/minimal.ndjson"),
                Paths.get("../testdata/src/rules.ndjson"),
                Paths.get("../testdata/src/unknown-parameter-type.ndjson")
        );
    }

    @ParameterizedTest
    @MethodSource("acceptance")
    void test(QueryTestCase testCase) throws IOException {
        ByteArrayOutputStream bytes = writeQueryResults(testCase, new ByteArrayOutputStream());
        String expected = new String(Files.readAllBytes(testCase.expected), UTF_8);
        String actual = new String(bytes.toByteArray(), UTF_8);
        assertThat(actual).isEqualTo(expected);
    }

    @ParameterizedTest
    @MethodSource("acceptance")
    @Disabled
    void updateExpectedFiles(QueryTestCase testCase) throws IOException {
        try (OutputStream out = Files.newOutputStream(testCase.expected)) {
            writeQueryResults(testCase, out);
        }
    }

    private static <T extends OutputStream> T writeQueryResults(QueryTestCase testCase, T out) throws IOException {
        Repository repository = createRepository();
        try (InputStream in = Files.newInputStream(testCase.source)) {
            try (NdjsonToMessageIterable envelopes = new NdjsonToMessageIterable(in, deserializer)) {
                envelopes.forEach(repository::update);
            }
        }
        Query query = new Query(repository);
        Object queryResults = testCase.query.apply(query);
        DefaultPrettyPrinter prettyPrinter = new DefaultPrettyPrinter()
                .withArrayIndenter(SYSTEM_LINEFEED_INSTANCE);
        OBJECT_MAPPER.writer(prettyPrinter).writeValue(out, queryResults);
        return out;
    }

    private static Repository createRepository() {
        return Repository.builder()
                .feature(INCLUDE_ATTACHMENTS, true)
                .feature(INCLUDE_STEP_DEFINITIONS, true)
                .feature(INCLUDE_SUGGESTIONS, true)
                .feature(INCLUDE_HOOKS, true)
                .feature(INCLUDE_GHERKIN_DOCUMENTS, true)
                .feature(INCLUDE_UNDEFINED_PARAMETER_TYPES, true)
                .build();
    }

    static Map<String, Function<Query, Object>> createQueries() {
        
        Map<String, Function<Query, Object>> queries = new LinkedHashMap<>();

        queries.put("countMostSevereTestStepResultStatus", Query::countMostSevereTestStepResultStatus);
        queries.put("countTestCasesStarted", Query::countTestCasesStarted);
        queries.put("findAllPickles", (query) -> query.findAllPickles().size());
        queries.put("findAllPickleSteps", (query) -> query.findAllPickleSteps().size());
        queries.put("findAllStepDefinitions", (query) -> query.findAllStepDefinitions().size());
        queries.put("findAllTestCaseStarted", (query) -> query.findAllTestCaseStarted().size());
        
        queries.put("findAllTestCaseStartedOrderBy", (query) -> query.findAllTestCaseStartedOrderBy(Query::findPickleBy, pickleComparator)
                .stream()
                .map(TestCaseStarted::getId)
                .collect(toList()));
        queries.put("findAllTestCaseFinished", (query) -> query.findAllTestCaseFinished().size());
        queries.put("findAllTestCaseFinishedOrderBy", (query) -> query.findAllTestCaseFinishedOrderBy(Query::findPickleBy, pickleComparator)
                .stream()
                .map(TestCaseFinished::getTestCaseStartedId)
                .collect(toList()));
        queries.put("findAllTestRunHookStarted", (query) -> query.findAllTestRunHookStarted().size());
        queries.put("findAllTestRunHookFinished", (query) -> query.findAllTestRunHookFinished().size());
        queries.put("findAllTestSteps", (query) -> query.findAllTestSteps().size());
        queries.put("findAllTestStepsStarted", (query) -> query.findAllTestStepStarted().size());
        queries.put("findAllTestStepsFinished", (query) -> query.findAllTestStepFinished().size());
        queries.put("findAllTestCases", (query) -> query.findAllTestCases().size());
        queries.put("findAllUndefinedParameterTypes", (query) -> query.findAllUndefinedParameterTypes().stream()
                .map(undefinedParameterType -> Arrays.asList(
                        undefinedParameterType.getName(),
                        undefinedParameterType.getExpression()
                ))
                .collect(toList()));

        queries.put("findAttachmentsBy", (query) -> {
            Map<String, Object> results = new LinkedHashMap<>();
            results.put("testStepFinished", query.findAllTestCaseStarted().stream()
                    .map(query::findTestStepFinishedAndTestStepBy)
                    .flatMap(Collection::stream)
                    .map(Map.Entry::getKey)
                    .map(query::findAttachmentsBy)
                    .flatMap(Collection::stream)
                    .map(attachment -> Arrays.asList(
                            attachment.getTestStepId(),
                            attachment.getTestCaseStartedId(),
                            attachment.getMediaType(),
                            attachment.getContentEncoding()
                    ))
                    .collect(toList()));
            results.put("testRunHookFinished", query.findAllTestRunHookFinished().stream()
                    .map(query::findAttachmentsBy)
                    .flatMap(Collection::stream)
                    .map(attachment -> Arrays.asList(
                            attachment.getTestRunHookStartedId(),
                            attachment.getMediaType(),
                            attachment.getContentEncoding()
                    ))
                    .collect(toList()));
            return results;
        });

        queries.put("findHookBy", (query) -> {
            Map<String, Object> results = new LinkedHashMap<>();
            results.put("testStep", query.findAllTestSteps().stream()
                    .map(query::findHookBy)
                    .map(hook -> hook.map(Hook::getId))
                    .filter(Optional::isPresent)
                    .collect(toList()));
            results.put("testRunHookStarted", query.findAllTestRunHookStarted().stream()
                    .map(query::findHookBy)
                    .map(hook -> hook.map(Hook::getId))
                    .filter(Optional::isPresent)
                    .collect(toList()));
            results.put("testRunHookFinished", query.findAllTestRunHookFinished().stream()
                    .map(query::findHookBy)
                    .map(hook -> hook.map(Hook::getId))
                    .filter(Optional::isPresent)
                    .collect(toList()));
            return results;
        });

        queries.put("findLineageBy", (query) -> {
            Map<String, Object> results = new LinkedHashMap<>();
            NamingStrategy namingStrategy = NamingStrategy.strategy(NamingStrategy.Strategy.LONG).build();
            results.put("testCaseStarted", query.findAllTestCaseStarted().stream()
                    .map(query::findLineageBy)
                    .filter(Optional::isPresent)
                    .map(Optional::get)
                    .map(namingStrategy::reduce)
                    .collect(toList()));
            results.put("testCaseFinished", query.findAllTestCaseFinished().stream()
                    .map(query::findLineageBy)
                    .filter(Optional::isPresent)
                    .map(Optional::get)
                    .map(namingStrategy::reduce)
                    .collect(toList()));
            results.put("pickle", query.findAllPickles().stream()
                    .map(query::findLineageBy)
                    .filter(Optional::isPresent)
                    .map(Optional::get)
                    .map(namingStrategy::reduce)
                    .collect(toList()));
            return results;
        });

        queries.put("findLocationOf", (query) -> query.findAllPickles().stream()
                .map(query::findLocationOf)
                .filter(Optional::isPresent)
                .collect(toList()));
        queries.put("findMeta", (query) -> query.findMeta().map(meta -> meta.getImplementation().getName()));

        queries.put("findMostSevereTestStepResultBy", (query) -> {
            Map<String, Object> results = new LinkedHashMap<>();
            results.put("testCaseStarted", query.findAllTestCaseStarted().stream()
                    .map(query::findMostSevereTestStepResultBy)
                    .filter(Optional::isPresent)
                    .map(Optional::get)
                    .map(TestStepResult::getStatus)
                    .collect(toList()));
            results.put("testCaseFinished", query.findAllTestCaseFinished().stream()
                    .map(query::findMostSevereTestStepResultBy)
                    .filter(Optional::isPresent)
                    .map(Optional::get)
                    .map(TestStepResult::getStatus)
                    .collect(toList()));
            return results;
        });

        queries.put("findPickleBy", (query) -> {
            Map<String, Object> results = new LinkedHashMap<>();
            results.put("testCaseStarted", query.findAllTestCaseStarted().stream()
                    .map(query::findPickleBy)
                    .map(pickle -> pickle.map(Pickle::getName))
                    .collect(toList()));
            results.put("testCaseFinished", query.findAllTestCaseFinished().stream()
                    .map(query::findPickleBy)
                    .map(pickle -> pickle.map(Pickle::getName))
                    .collect(toList()));
            results.put("testStepStarted", query.findAllTestStepStarted().stream()
                    .map(query::findPickleBy)
                    .map(pickle -> pickle.map(Pickle::getName))
                    .collect(toList()));
            results.put("testStepFinished", query.findAllTestStepFinished().stream()
                    .map(query::findPickleBy)
                    .map(pickle -> pickle.map(Pickle::getName))
                    .collect(toList()));
            return results;
        });

        queries.put("findPickleStepBy", (query) -> query.findAllTestSteps().stream()
                .map(query::findPickleStepBy)
                .map(pickleStep -> pickleStep.map(PickleStep::getText))
                .filter(Optional::isPresent)
                .collect(toList()));
        queries.put("findStepBy", (query) -> query.findAllPickleSteps().stream()
                .map(query::findStepBy)
                .map(step -> step.map(Step::getText))
                .collect(toList()));
        queries.put("findStepDefinitionsBy", (query) -> query.findAllTestSteps().stream()
                .map(query::findStepDefinitionsBy)
                .map(stepDefinitions -> stepDefinitions.stream().map(StepDefinition::getId)
                        .collect(toList()))
                .collect(toList()));

        queries.put("findSuggestionsBy", (query) -> {
            Map<String, Object> results = new LinkedHashMap<>();
            results.put("pickleStep", query.findAllPickleSteps().stream()
                    .map(query::findSuggestionsBy)
                    .flatMap(Collection::stream)
                    .map(Suggestion::getId)
                    .collect(toList()));
            results.put("pickle", query.findAllPickles().stream()
                    .map(query::findSuggestionsBy)
                    .flatMap(Collection::stream)
                    .map(Suggestion::getId)
                    .collect(toList()));
            return results;
        });

        queries.put("findUnambiguousStepDefinitionBy", (query) -> query.findAllTestSteps().stream()
                .map(query::findUnambiguousStepDefinitionBy)
                .filter(Optional::isPresent)
                .map(Optional::get)
                .map(StepDefinition::getId));

        queries.put("findTestCaseStartedBy", (query) -> {
            Map<String, Object> results = new LinkedHashMap<>();
            results.put("testCaseFinished", query.findAllTestCaseFinished().stream()
                    .map(query::findTestCaseStartedBy)
                    .map(testCase -> testCase.map(TestCaseStarted::getId))
                    .collect(toList()));
            results.put("testStepStarted", query.findAllTestStepStarted().stream()
                    .map(query::findTestCaseStartedBy)
                    .map(testCase -> testCase.map(TestCaseStarted::getId))
                    .collect(toList()));
            results.put("testStepFinished", query.findAllTestStepFinished().stream()
                    .map(query::findTestCaseStartedBy)
                    .map(testCase -> testCase.map(TestCaseStarted::getId))
                    .collect(toList()));
            return results;
        });

        queries.put("findTestCaseBy", (query) -> {
            Map<String, Object> results = new LinkedHashMap<>();
            results.put("testCaseStarted", query.findAllTestCaseStarted().stream()
                    .map(query::findTestCaseBy)
                    .map(testCase -> testCase.map(TestCase::getId))
                    .collect(toList()));
            results.put("testCaseFinished", query.findAllTestCaseFinished().stream()
                    .map(query::findTestCaseBy)
                    .map(testCase -> testCase.map(TestCase::getId))
                    .collect(toList()));
            results.put("testStepStarted", query.findAllTestStepStarted().stream()
                    .map(query::findTestCaseBy)
                    .map(testCase -> testCase.map(TestCase::getId))
                    .collect(toList()));
            results.put("testStepFinished", query.findAllTestStepFinished().stream()
                    .map(query::findTestCaseBy)
                    .map(testCase -> testCase.map(TestCase::getId))
                    .collect(toList()));
            return results;
        });

        queries.put("findTestCaseDurationBy", (query) -> {
            Map<String, Object> results = new LinkedHashMap<>();
            results.put("testCaseStarted", query.findAllTestCaseStarted().stream()
                    .map(query::findTestCaseDurationBy)
                    .map(duration -> duration.map(Convertor::toMessage))
                    .collect(toList()));
            results.put("testCaseFinished", query.findAllTestCaseFinished().stream()
                    .map(query::findTestCaseDurationBy)
                    .map(duration -> duration.map(Convertor::toMessage))
                    .collect(toList()));
            return results;
        });

        queries.put("findTestCaseFinishedBy", (query) -> query.findAllTestCaseStarted().stream()
                .map(query::findTestCaseFinishedBy)
                .map(testCaseFinished -> testCaseFinished.map(TestCaseFinished::getTestCaseStartedId))
                .collect(toList()));
        queries.put("findTestRunDuration", (query) -> query.findTestRunDuration()
                .map(Convertor::toMessage));
        queries.put("findTestRunFinished", Query::findTestRunFinished);
        queries.put("findTestRunStarted", Query::findTestRunStarted);
        queries.put("findTestStepBy", (query) -> query.findAllTestCaseStarted().stream()
                .map(query::findTestStepsStartedBy)
                .flatMap(Collection::stream)
                .map(query::findTestStepBy)
                .map(testStep -> testStep.map(TestStep::getId))
                .collect(toList()));

        queries.put("findTestStepsStartedBy", (query) -> {
            Map<String, Object> results = new LinkedHashMap<>();

            results.put("testCaseStarted", query.findAllTestCaseStarted().stream()
                    .map(query::findTestStepsStartedBy)
                    .map(Collection::stream)
                    .map(testStepStarted -> testStepStarted.map(TestStepStarted::getTestStepId))
                    .collect(toList()));
            results.put("testCaseFinished", query.findAllTestCaseFinished().stream()
                    .map(query::findTestStepsStartedBy)
                    .map(Collection::stream)
                    .map(testStepStarted -> testStepStarted.map(TestStepStarted::getTestStepId))
                    .collect(toList()));
            return results;
        });

        queries.put("findTestRunHookFinishedBy", (query) -> query.findAllTestRunHookStarted().stream()
                .map(query::findTestRunHookFinishedBy)
                .map(testRunHookFinished -> testRunHookFinished.map(TestRunHookFinished::getTestRunHookStartedId))
                .collect(toList()));
        queries.put("findTestRunHookStartedBy", (query) -> query.findAllTestRunHookFinished().stream()
                .map(query::findTestRunHookStartedBy)
                .map(testRunHookStarted -> testRunHookStarted.map(TestRunHookStarted::getId))
                .collect(toList()));

        queries.put("findTestStepByTestStepFinished", (query) -> {
            Map<String, Object> results = new LinkedHashMap<>();

            results.put("testCaseStarted", query.findAllTestCaseStarted().stream()
                    .map(query::findTestStepsFinishedBy)
                    .flatMap(Collection::stream)
                    .map(query::findTestStepBy)
                    .map(testStep -> testStep.map(TestStep::getId))
                    .collect(toList()));
            results.put("testCaseFinished", query.findAllTestCaseFinished().stream()
                    .map(query::findTestStepsFinishedBy)
                    .flatMap(Collection::stream)
                    .map(query::findTestStepBy)
                    .map(testStep -> testStep.map(TestStep::getId))
                    .collect(toList()));

            return results;
        });

        queries.put("findTestStepsFinishedBy", (query) -> query.findAllTestCaseStarted().stream()
                .map(query::findTestStepsFinishedBy)
                .map(testStepFinisheds -> testStepFinisheds.stream().map(TestStepFinished::getTestStepId).collect(toList()))
                .collect(toList()));
        queries.put("findTestStepFinishedAndTestStepBy", (query) -> query.findAllTestCaseStarted().stream()
                .map(query::findTestStepFinishedAndTestStepBy)
                .flatMap(Collection::stream)
                .map(entry -> asList(entry.getKey().getTestStepId(), entry.getValue().getId()))
                .collect(toList()));

        return queries;
    }

    static class QueryTestCase {
        private final String methodName;
        private final String name;
        private final Path source;
        private final Path expected;
        private final Function<Query, Object> query;

        QueryTestCase(String methodName, Path source, Function<Query, Object> query) {
            this.methodName = methodName;
            this.source = source;
            this.query = query;
            String fileName = source.getFileName().toString();
            this.name = fileName.substring(0, fileName.lastIndexOf(".ndjson"));
            this.expected = source.getParent().resolve(name + "." + methodName + ".results.json");
        }

        @Override
        public String toString() {
            return name + " -> " + methodName;
        }
    }

}

File: dotnet/QueryTest/QueryAcceptanceTest.cs
using FluentAssertions;
using Io.Cucumber.Messages.Types;
using Io.Cucumber.Query;
using System.Reflection;
using System.Text;
using System.Text.Json;
using System.Text.Json.Nodes;

namespace QueryTest
{
    [TestClass]
    public class QueryAcceptanceTest
    {
        private static readonly string[] _sources = new[]
        {
            "attachments.ndjson",
            "empty.ndjson",
            "examples-tables.ndjson",
            "global-hooks.ndjson",
            "global-hooks-attachments.ndjson",
            "hooks.ndjson",
            "minimal.ndjson",
            "rules.ndjson",
            "unknown-parameter-type.ndjson"
        };

        public static IEnumerable<object[]> Acceptance()
        {
            var queries = createQueries();
            foreach (var file in _sources)
            {
                foreach (var query in queries)
                {
                    yield return new object[] { new QueryTestCase(query.Key, file, query.Value) };
                }
            }
        }

        private static Dictionary<string, Func<Query, object>> createQueries()
        {
            var namingStrategy = NamingStrategy.Create(NamingStrategy.Strategy.LONG);
            var queries = new Dictionary<string, Func<Query, object>>
            {
                ["countMostSevereTestStepResultStatus"] = q => q.CountMostSevereTestStepResultStatus()
                                                                .ToDictionary(
                                                                    kvp => kvp.Key.EnumDescription<TestStepResultStatus>(),
                                                                    kvp => kvp.Value),
                ["countTestCasesStarted"] = q => q.TestCasesStartedCount,
                ["findAllPickles"] = q => q.FindAllPickles().Count(),
                ["findAllPickleSteps"] = q => q.FindAllPickleSteps().Count(),
                ["findAllStepDefinitions"] = q => q.FindAllStepDefinitions().Count(),
                ["findAllTestCaseStarted"] = q => q.FindAllTestCaseStarted().Count(),
                ["findAllTestCaseFinished"] = q => q.FindAllTestCaseFinished().Count(),
                ["findAllTestRunHookStarted"] = q => q.FindAllTestRunHookStarted().Count(),
                ["findAllTestRunHookFinished"] = q => q.FindAllTestRunHookFinished().Count(),
                ["findAllTestSteps"] = q => q.FindAllTestSteps().Count(),
                ["findAllTestStepsStarted"] = q => q.FindAllTestStepStarted().Count(),
                ["findAllTestStepsFinished"] = q => q.FindAllTestStepFinished().Count(),
                ["findAllTestCases"] = q => q.FindAllTestCases().Count(),
                ["findAllUndefinedParameterTypes"] = q => q.FindAllUndefinedParameterTypes()
                    .Select(upt => new object?[]
                        {
                            upt.Name,
                            upt.Expression
                        }).ToList(),
                ["findAttachmentsBy"] = q => new Dictionary<string, object>
                {
                    ["testStepFinished"] = q.FindAllTestCaseStarted()
                        .SelectMany(tcs => q.FindTestStepFinishedAndTestStepBy(tcs))
                        .Select(pair => pair.Item1)
                        .SelectMany(tsf => q.FindAttachmentsBy(tsf))
                        .Select(att => new object?[]
                        {
                            att.TestStepId,
                            att.TestCaseStartedId,
                            att.MediaType,
                            att.ContentEncoding
                        }).ToList(),
                    ["testRunHookFinished"] = q.FindAllTestRunHookFinished()
                        .SelectMany(trhf => q.FindAttachmentsBy(trhf))
                        .Select(att => new object?[]
                        {
                            att.TestRunHookStartedId,
                            att.MediaType,
                            att.ContentEncoding
                        }).ToList()
                },
                ["findHookBy"] = q => new Dictionary<string, object>
                {
                    ["testStep"] = q.FindAllTestSteps()
                        .Select(ts => q.FindHookBy(ts)?.Id)
                        .Where(id => id != null)
                        .ToList(),
                    ["testRunHookStarted"] = q.FindAllTestRunHookStarted()
                        .Select(trhs => q.FindHookBy(trhs)?.Id)
                        .Where(id => id != null)
                        .ToList(),
                    ["testRunHookFinished"] = q.FindAllTestRunHookFinished()
                        .Select(trhf => q.FindHookBy(trhf)?.Id)
                        .Where(id => id != null)
                        .ToList(),
                },
                ["findLineageBy"] = q => new Dictionary<string, object>
                {
                    ["testCaseStarted"] = q.FindAllTestCaseStarted()
                        .Select(tcs => q.FindLineageBy(tcs))
                        .Where(lineage => lineage != null)
                        .Select(lineage => namingStrategy.Reduce(lineage))
                        .ToList(),
                    ["testCaseFinished"] = q.FindAllTestCaseFinished()
                        .Select(tcs => q.FindLineageBy(tcs))
                        .Where(lineage => lineage != null)
                        .Select(lineage => namingStrategy.Reduce(lineage))
                        .ToList(),
                    ["pickle"] = q.FindAllPickles()
                        .Select(pickle => q.FindLineageBy(pickle))
                        .Where(lineage => lineage != null)
                        .Select(lineage => namingStrategy.Reduce(lineage))
                        .ToList()
                },
                ["findLocationOf"] = q => q.FindAllPickles()
                    .Select(pickle => q.FindLocationOf(pickle))
                    .Where(loc => loc != null)
                    .ToList(),
                ["findMeta"] = q => q.FindMeta()?.Implementation?.Name!,
                ["findMostSevereTestStepResultBy"] = q => new Dictionary<string, object>
                {
                    ["testCaseStarted"] = q.FindAllTestCaseStarted()
                        .Select(tcs => q.FindMostSevereTestStepResultBy(tcs)?.Status)
                        .Where(status => status != null)
                        .ToList(),
                    ["testCaseFinished"] = q.FindAllTestCaseFinished()
                        .Select(tcf => q.FindMostSevereTestStepResultBy(tcf)?.Status)
                        .Where(status => status != null)
                        .ToList()
                },
                ["findPickleBy"] = q => new Dictionary<string, object>
                {
                    ["testCaseStarted"] = q.FindAllTestCaseStarted()
                        .Select(tcs => q.FindPickleBy(tcs)?.Name)
                        .ToList(),
                    ["testCaseFinished"] = q.FindAllTestCaseFinished()
                        .Select(tcf => q.FindPickleBy(tcf)?.Name)
                        .ToList(),
                    ["testStepStarted"] = q.FindAllTestStepStarted()
                        .Select(tss => q.FindPickleBy(tss)?.Name)
                        .ToList(),
                    ["testStepFinished"] = q.FindAllTestStepFinished()
                        .Select(tsf => q.FindPickleBy(tsf)?.Name)
                        .ToList()
                },
                ["findPickleStepBy"] = q => q.FindAllTestSteps()
                    .Select(ts => q.FindPickleStepBy(ts)?.Text)
                    .Where(text => text != null)
                    .ToList(),
                ["findStepBy"] = q => q.FindAllPickleSteps()
                    .Select(ps => q.FindStepBy(ps)?.Text)
                    .ToList(),
                ["findStepDefinitionsBy"] = q => q.FindAllTestSteps()
                    .Select(ts => q.FindStepDefinitionsBy(ts).Select(sd => sd.Id).ToList())
                    .ToList(),
                ["findSuggestionsBy"] = q => new Dictionary<string, object>
                {
                    ["pickleStep"] = q.FindAllPickleSteps()
                        .SelectMany(ps => q.FindSuggestionsBy(ps))
                        .Select(s => s.Id)
                        .ToList(),
                    ["pickle"] = q.FindAllPickles()
                        .SelectMany(pickle => q.FindSuggestionsBy(pickle))
                        .Select(s => s.Id)
                        .ToList()
                },
                ["findUnambiguousStepDefinitionBy"] = q => q.FindAllTestSteps()
                    .Select(ts => q.FindUnambiguousStepDefinitionBy(ts)?.Id)
                    .Where(id => id != null)
                    .ToList(),
                ["findTestCaseStartedBy"] = q => new Dictionary<string, object>
                {
                    ["testCaseFinished"] = q.FindAllTestCaseFinished()
                        .Select(tcf => q.FindTestCaseStartedBy(tcf)?.Id)
                        .ToList(),
                    ["testStepStarted"] = q.FindAllTestStepStarted()
                        .Select(tss => q.FindTestCaseStartedBy(tss)?.Id)
                        .ToList(),
                    ["testStepFinished"] = q.FindAllTestStepFinished()
                        .Select(tsf => q.FindTestCaseStartedBy(tsf)?.Id)
                        .ToList()
                },
                ["findTestCaseBy"] = q => new Dictionary<string, object>
                {
                    ["testCaseStarted"] = q.FindAllTestCaseStarted()
                        .Select(tcs => q.FindTestCaseBy(tcs)?.Id)
                        .ToList(),
                    ["testCaseFinished"] = q.FindAllTestCaseFinished()
                        .Select(tcf => q.FindTestCaseBy(tcf)?.Id)
                        .ToList(),
                    ["testStepStarted"] = q.FindAllTestStepStarted()
                        .Select(tss => q.FindTestCaseBy(tss)?.Id)
                        .ToList(),
                    ["testStepFinished"] = q.FindAllTestStepFinished()
                        .Select(tsf => q.FindTestCaseBy(tsf)?.Id)
                        .ToList()
                },
                ["findTestCaseDurationBy"] = q => new Dictionary<string, object>
                {
                    ["testCaseStarted"] = q.FindAllTestCaseStarted()
                        .Select(tcs => ConvertTimeSpanToTimestamp(q.FindTestCaseDurationBy(tcs)))
                        .ToList(),
                    ["testCaseFinished"] = q.FindAllTestCaseFinished()
                        .Select(tcf => ConvertTimeSpanToTimestamp(q.FindTestCaseDurationBy(tcf)))
                        .ToList()
                },
                ["findTestCaseFinishedBy"] = q => q.FindAllTestCaseStarted()
                    .Select(tcs => q.FindTestCaseFinishedBy(tcs)?.TestCaseStartedId)
                    .ToList(),
                ["findTestRunDuration"] = q => ConvertTimeSpanToTimestamp(q.FindTestRunDuration())!,
                ["findTestRunFinished"] = q => q.FindTestRunFinished()!,
                ["findTestRunStarted"] = q => q.FindTestRunStarted()!,
                ["findTestStepBy"] = q => q.FindAllTestCaseStarted()
                    .SelectMany(tcs => q.FindTestStepsStartedBy(tcs))
                    .Select(tss => q.FindTestStepBy(tss)?.Id)
                    .ToList(),
                ["findTestStepsStartedBy"] = q => new Dictionary<string, object>
                {
                    ["testCaseStarted"] = q.FindAllTestCaseStarted()
                        .Select(tcs => q.FindTestStepsStartedBy(tcs).Select(tss => tss.TestStepId).ToList())
                        .ToList(),
                    ["testCaseFinished"] = q.FindAllTestCaseFinished()
                        .Select(tcf => q.FindTestStepsStartedBy(tcf).Select(tss => tss.TestStepId).ToList())
                        .ToList()
                },
                ["findTestRunHookFinishedBy"] = q => q.FindAllTestRunHookStarted()
                    .Select(trhs => q.FindTestRunHookFinishedBy(trhs)?.TestRunHookStartedId)
                    .ToList(),
                ["findTestRunHookStartedBy"] = q => q.FindAllTestRunHookFinished()
                    .Select(trhf => q.FindTestRunHookStartedBy(trhf)?.Id)
                    .ToList(),
                ["findTestStepByTestStepFinished"] = q => new Dictionary<string, object>
                {
                    ["testCaseStarted"] = q.FindAllTestCaseStarted()
                        .SelectMany(tcs => q.FindTestStepsFinishedBy(tcs))
                        .Select(tsf => q.FindTestStepBy(tsf)?.Id)
                        .ToList(),
                    ["testCaseFinished"] = q.FindAllTestCaseFinished()
                        .SelectMany(tcf => q.FindTestStepsFinishedBy(tcf))
                        .Select(tsf => q.FindTestStepBy(tsf)?.Id)
                        .ToList()
                },
                ["findTestStepsFinishedBy"] = q => q.FindAllTestCaseStarted()
                    .Select(tcs => q.FindTestStepsFinishedBy(tcs).Select(tsf => tsf.TestStepId).ToList())
                    .ToList(),
                ["findTestStepFinishedAndTestStepBy"] = q => q.FindAllTestCaseStarted()
                    .SelectMany(tcs => q.FindTestStepFinishedAndTestStepBy(tcs))
                    .Select(pair => new object?[] { pair.Item1.TestStepId, pair.Item2.Id })
                    .ToList()
            };
            return queries;
        }

        [TestMethod]
        [DynamicData(nameof(Acceptance), DynamicDataSourceType.Method)]
        public void Test(QueryTestCase testCase)
        {
            var actual = WriteQueryResults(testCase);
            var expected = ReadResourceAsString(testCase.ExpectedResourceName);

            // Compare as JSON for robust diff
            var actualJson = JsonNode.Parse(actual);
            var expectedJson = JsonNode.Parse(expected);

            actualJson!.ToJsonString().Should().Be(expectedJson!.ToJsonString(),
                $"Query results for {testCase.Name}.{testCase.MethodName} do not match expected results.");
        }

        private static string WriteQueryResults(QueryTestCase testCase)
        {
            using var inStream = ReadResourceAsStream(testCase.SourceResourceName);
            using var reader = new StreamReader(inStream, Encoding.UTF8);

            var repository = CreateRepository();

            // Read NDJSON lines and update _query
            string? line;
            while ((line = reader.ReadLine()) != null)
            {
                if (string.IsNullOrWhiteSpace(line)) continue;
                var envelope = NdjsonSerializer.Deserialize<Envelope>(line);
                repository.Update(envelope);
            }

            var query = new Query(repository);
            var queryResults = testCase.Query(query);
            var options = new JsonSerializerOptions(NdjsonSerializer.JsonOptions);
            options.Converters.Add(new TimestampOrderedConverter());
            options.Converters.Add(new TestRunStartedOrderedConverter());
            options.Converters.Add(new TestRunFinishedOrderedConverter());

            return JsonSerializer.Serialize(queryResults, options);
        }

        private static Repository CreateRepository()
        {
            return Repository.CreateWithAllFeatures();
        }

        private static Timestamp? ConvertTimeSpanToTimestamp(TimeSpan? duration)
        {
            if (duration == null) return null;
            return new Timestamp(
                (long)duration.Value.TotalSeconds,
                (int)((duration.Value.Ticks % TimeSpan.TicksPerSecond) * 100)
            );
        }

        private static Stream ReadResourceAsStream(string resourceName)
        {
            var fullName = Path.Combine(TestFolderHelper.TestFolder, "src", resourceName);
            if (!File.Exists(fullName))
                throw new FileNotFoundException($"Resource {fullName} not found.");
            return File.OpenRead(fullName);
        }

        private static string ReadResourceAsString(string resourceName)
        {
            using var stream = ReadResourceAsStream(resourceName);
            using var reader = new StreamReader(stream, Encoding.UTF8);
            return reader.ReadToEnd();
        }

        public class QueryTestCase
        {
            public string SourceResourceName { get; }
            public string ExpectedResourceName { get; }
            public string Name { get; }
            public string MethodName { get; }
            public Func<Query, object> Query { get; }

            public QueryTestCase(string methodName, string source, Func<Query, object> query)
            {
                MethodName = methodName;
                Name = source.Substring(0, source.LastIndexOf(".ndjson", StringComparison.Ordinal));
                SourceResourceName = source;
                ExpectedResourceName = $"{Name}.{MethodName}.results.json";
                Query = query;
            }

            public override string ToString() => Name;
        }
    }
}